pkg/pmem-csi-driver/pmem-csi-driver.go

```diff
@@ -184,7 +184,7 @@ func (pmemd *pmemDriver) Run() error {
 			if err := pmemd.registerNodeController(); err != nil {
 				return err
 			}
-			if err := s.Start(pmemd.cfg.Endpoint, nil, ids, ns); err != nil {
+			if err := s.Start(pmemd.cfg.Endpoint, nil, ids, ns, cs /* cs only needed for testing */); err != nil {
```

Comment:
@avalluri, @okartau: what do you think about this?

It exposes the controller implementation also on a node's csi.sock where it can be accessed for testing purposes, using the already existing mechanism (socat socket forwarding).

It's currently done unconditional. There's no harm in that (anyone who can connect to csi.sock can do all kinds of havoc, whether this particular interface is there or not), but it might be seen as exposing a debug interface also in production.

I could make this conditional and only enable it in our testing deployment. I'm just not sure whether that's worth the effort.

Reply:
@pohly I personally not favor of exposing interfaces that are used just for tests. In this particular case, I would prefer testcase that checks if the existing volume is "publishable" after node/cluster reboot and avoid ListVolulems check.

Reply:
I also think this (exposing of test features) should be avoided if we can.

Reply:
What if it is optional? Is that acceptable?

Exposing it enables writing tests that are more focused on a particular aspect and thus easier to debug when something goes wrong. I think that's important.

Testing cache volumes will also be easier that way.

Reply:
I think it's acceptable if test/debug feature can be turned on/off during build time, with default=off in production version.

Reply:
> What if it is optional? Is that acceptable?
> 
"Optional" you mean exposing "cs" only in test builds?, that's fine for me.

Reply:
I was thinking of doing it via a runtime command line parameter. This is conceptually similar with having debug output built into production binaries.

Reply:
Please check whether 4d8ee6f0fd12b2b6df7ec778fa3025b643b8b07a addresses this concern.

Reply:
> > Exposing it enables writing tests that are more focused on a particular aspect and thus easier to debug when something goes wrong. I think that's important.
> > Testing cache volumes will also be easier that way.
> 
> To demonstrate what I meant with that I added a test for cache volumes: [cd25587](https://github.com/intel/pmem-CSI/commit/cd25587ef1e00dd34e954f303512f4cee449d958)

@pohly Why tests are focusing on volume lists on a node, instead wouldn't it be good enough if we check that the volume could be publishable on the node where the volume expected to be created.

Reply:
> Why tests are focusing on volume lists on a node, instead wouldn't it be good enough if we check that the volume could be publishable on the node where the volume expected to be created.

Because that makes tests more complicated and harder to debug when they fail.